runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,224 @@
+                    "subscribe: Max " + MAX_SUBSCRIBERS + " hit! Consider consolidating "
+                            + streamListener);
+        }
+        if (subscriptions.containsKey(streamListener)) {
+            throw new StreamSubscriptionException(
+                    "StreamManager:subscriber already registered "
+                            + streamListener + ". Maybe call unregister first?");
+        }
+        log.info("StreamManager::subscribe {}, startAddress {}, namespace {}, tables {}",
+                streamListener, startAddress, namespace, tablesOfInterest.toString());
+        SubscriptionThread st = new SubscriptionThread(this, streamListener,
+                namespace, tablesOfInterest, startAddress);
+        subscriptions.put(streamListener, st);
+        st.start();
+    }
+
+    /**
+     * Unsubscribe a prior subscription.
+     *
+     * @param streamListener Client listener.
+     */
+    public synchronized void unsubscribe(@Nonnull StreamListener streamListener) {
+        SubscriptionThread st = subscriptions.remove(streamListener);
+        if (st == null) {
+            log.warn("StreamManager::unsubscribe has no context for {}", streamListener.toString());
+            return;
+        }
+        st.shutdown.set(true);
+        st.interrupt();
+        log.info("Unsubscribed StreamListener {}", streamListener.toString());
+    }
+
+    class SubscriptionThread extends Thread {
+        /**
+         * The callback context
+         */
+        private final StreamListener listener;
+
+        /**
+         * Parent class to manage errors.
+         */
+        @Getter
+        private final StreamManager streamManager;
+
+        /**
+         * Namespace of interest.
+         */
+        @Getter
+        private final String namespace;
+
+        /**
+         * Tables of interest.
+         *
+         * Map of (Stream Id - TableSchema)
+         */
+        @Getter
+        private final Map<UUID,
+                TableSchema<? extends Message, ? extends Message, ? extends Message>> tablesOfInterest;
+
+        /**
+         * Starting address.
+         */
+        private final long startAddress;
+
+        public AtomicBoolean shutdown;
+
+        /**
+         * A reference to the underlying stream
+         */
+        @Getter
+        private final IStreamView txnStream;
+
+        public void run() {
+            try {
+                log.info("{}:Seeking txStream to {}", listener.toString(), startAddress + 1);
+                txnStream.seek(startAddress + 1);
+            } catch (Throwable throwable) {
+                log.warn("{}:onError at seek to {}", listener.toString(), startAddress + 1);
+                listener.onError(throwable);
+                streamManager.unsubscribe(listener);
+                return;
+            }
+
+            try {
+                while (!shutdown.get()) {
+                    while (!txnStream.hasNext()) {
```

Comment:
I wouldn't use this API. It's not used and might be buggy. I recommend that you just try to sync anyways. If there are no deltas, the callbacks won't trigger.

Reply:
ok - will choose to sleep if there are no deltas. btw we should consider either fixing this buggy api or removing it 
I see it used in other places as well.

Reply:
`AbstractContextStreamView::hasNext` is actually not being used by the VLO, or internal streaming that's why I'm recommending not to use it, we just dont know it it works. Maybe you can just remove it from the stream API itself.